src/Orleans.ScheduledJobs/LocalScheduledJobManager.cs

```diff
@@ -0,0 +1,473 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Orleans.Hosting;
+using Orleans.Runtime;
+using Orleans.Runtime.Internal;
+using Orleans.Runtime.Scheduler;
+
+namespace Orleans.ScheduledJobs;
+
+/// <summary>
+/// Provides functionality for scheduling and managing jobs on the local silo.
+/// </summary>
+public interface ILocalScheduledJobManager
+{
+    /// <summary>
+    /// Schedules a job to be executed at a specific time on the target grain.
+    /// </summary>
+    /// <param name="target">The grain identifier of the target grain that will receive the scheduled job.</param>
+    /// <param name="jobName">The name of the job for identification purposes.</param>
+    /// <param name="dueTime">The date and time when the job should be executed.</param>
+    /// <param name="metadata">Optional metadata associated with the job.</param>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation that returns the scheduled job.</returns>
+    Task<IScheduledJob> ScheduleJobAsync(GrainId target, string jobName, DateTimeOffset dueTime, IReadOnlyDictionary<string, string>? metadata = null);
+
+    /// <summary>
+    /// Attempts to cancel a previously scheduled job.
+    /// </summary>
+    /// <param name="job">The scheduled job to cancel.</param>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation that returns <see langword="true"/> if the job was successfully canceled; otherwise, <see langword="false"/>.</returns>
+    Task<bool> TryCancelScheduledJobAsync(IScheduledJob job);
+}
+
+/// <inheritdoc/>
+internal partial class LocalScheduledJobManager : SystemTarget, ILocalScheduledJobManager, ILifecycleParticipant<ISiloLifecycle>
+{
+    private readonly JobShardManager _shardManager;
+    private readonly IAsyncEnumerable<ClusterMembershipSnapshot> _clusterMembershipUpdates;
+    private readonly ILogger<LocalScheduledJobManager> _logger;
+    private readonly ScheduledJobsOptions _options;
+    private CancellationTokenSource _cts = new();
+    private Task? _listenForClusterChangesTask = null;
+    private Task? _watchForShardtoStartTask = null;
+    private readonly ConcurrentDictionary<DateTimeOffset, ConcurrentDictionary<string, JobShard>> _shardCache = new();
+    private readonly ConcurrentDictionary<string, Task> _runningShards = new();
+    private readonly int MaxJobCountPerShard = 1000;
+    private readonly SemaphoreSlim _jobConcurrencyLimiter;
+
+    private static readonly IDictionary<string, string> EmptyMetadata = new Dictionary<string, string>();
+
+    public LocalScheduledJobManager(
+        JobShardManager shardManager,
+        IClusterMembershipService clusterMembership,
+        IOptions<ScheduledJobsOptions> options,
+        SystemTargetShared shared,
+        ILogger<LocalScheduledJobManager> logger)
+        : base(SystemTargetGrainId.CreateGrainType("scheduledjobs-manager"), shared)
```

Comment:
"job-manager" and consider renaming the class, too. We could drop Scheduled from all nouns and just make it "Jobs"